lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,243 @@
+require 'set'
+require 'yaml'
+require 'open-uri'
+require 'objspace'
+require 'fog/core'
+
+module Fog
+  module Orchestration
+    module Util
+      #
+      # Resolve get_file resources found in a HOT template populating
+      #  a files Hash conforming to Heat Specs
+      #  https://developer.openstack.org/api-ref/orchestration/v1/index.html?expanded=create-stack-detail#stacks
+      #
+      # Files present in :files are not processed further. The others
+      #   are added to the Hash. This behavior is the same implemented in openstack-infra/shade
+      #   see https://github.com/openstack-infra/shade/blob/1d16f64fbf376a956cafed1b3edd8e51ccc16f2c/shade/openstackcloud.py#L1200
+
+      #
+      # This implementation just process nested templates but not resource
+      #  registries.
+      class RecursiveHotFileLoader
+        attr_reader :files
+        attr_reader :template
+        attr_reader :template_ori
+        attr_reader :visited
+        attr_reader :max_files_size
+
+        def initialize(template, files = nil)
+          @template = template
+          @template_ori = template
+          @template_base_url = nil
+          @files = files || {}
```

Comment:
My original recommendation was that if `files` are already given by caller, don't go through this `RecusiveHotFileLoader` at all. My assumption is that the user already resolved `files`, there is no need to read and process the template.

You decided to accept `files` as an argument and always go through `RecusiveHotFileLoader`, which is time consuming but may be useful one one scenario - if `files` is incomplete, and we still want to help discover the rest. But in your code I don't see you treat `files` as a user passed in argument.

We don't need to be this flexible. I would still recommend what I originally recommended.

Reply:
@bzwei on this subject, I implemented the same behavior of heatclient so openstack user can expect a similar behavior between CFME and other python tools (Ansible Tower).

I'm moving that code in a different patch.